opm/simulators/linalg/DILU.hpp

```diff
@@ -17,90 +17,109 @@
 #ifndef OPM_DILU_HEADER_INCLUDED
 #define OPM_DILU_HEADER_INCLUDED
 
-#include <config.h>
 #include <opm/common/ErrorMacros.hpp>
 #include <opm/common/TimingMacros.hpp>
 #include <opm/simulators/linalg/PreconditionerWithUpdate.hpp>
 
 #include <dune/common/fmatrix.hh>
-#include <dune/common/version.hh>
 #include <dune/common/unused.hh>
+#include <dune/common/version.hh>
 #include <dune/istl/bcrsmatrix.hh>
 
+#include <opm/simulators/linalg/GraphColoring.hpp>
+
+#include <cstddef>
+#include <vector>
+
+#if HAVE_OPENMP
+#include <omp.h>
+#endif
 
+// TODO: rewrite factory and constructor to keep track of a number of threads variable
 namespace Dune
 {
 
-/*! \brief The sequential DILU preconditioner.
+/*! \brief The OpenMP thread parallelized DILU preconditioner.
+ *  \details Safe to run serially without OpenMP. When run in parallel
+             the matrix is assumed to be symmetric.
 
    \tparam M The matrix type to operate on
    \tparam X Type of the update
    \tparam Y Type of the defect
 */
 template <class M, class X, class Y>
-class SeqDilu : public PreconditionerWithUpdate<X, Y>
+class MultithreadDILU : public PreconditionerWithUpdate<X, Y>
 {
-    public:
-        //! \brief The matrix type the preconditioner is for.
-        using matrix_type = M;
-        //! \brief The domain type of the preconditioner.
-        using domain_type = X;
-        //! \brief The range type of the preconditioner.
-        using range_type = Y;
-        //! \brief The field type of the preconditioner.
-        using field_type = typename X::field_type;
-        //! \brief scalar type underlying the field_type
-
-    /*! \brief Constructor.
-       Constructor gets all parameters to operate the prec.
+public:
+    //! \brief The matrix type the preconditioner is for.
+    using matrix_type = M;
+    //! \brief The domain type of the preconditioner.
+    using domain_type = X;
+    //! \brief The range type of the preconditioner.
+    using range_type = Y;
+    //! \brief The field type of the preconditioner.
+    using field_type = typename X::field_type;
+    //! \brief scalar type underlying the field_type
+
+    /*! \brief Constructor gets all parameters to operate the prec.
        \param A The matrix to operate on.
     */
-    SeqDilu(const M& A)
+    MultithreadDILU(const M& A)
         : A_(A)
+        , A_reordered_(M(A_.N(), A_.N(), A_.nonzeroes(), M::row_wise))
```

Comment:
Not quite sure how to avoid this. Naively removing that statement and putting "A_reordered_ = M(A_.N(), A_.N(), A_.nonzeroes(), M::row_wise);" inside the if using_multithreading yields "InvalidStateException [operator=:/usr/include/dune/istl/bcrsmatrix.hh:915]: BCRSMatrix can only be copied when both target and source are empty or fully built)". Is there some clever way of initializing in this variable inside constructor without using the = operator?

Reply:
make A_reordered_ a std::optional<M>. This way you can delay initializing it the if() below, i.e.
```
A_ordered_ = M(A_.N(), A_.N(), A_.nonzeroes(), M::row_wise);
```
Finally you then have to use pointer-like semantics for A_reordered_ where you use it.

Reply:
> ```C++
> A_reordered_ = M(A_.N(), A_.N(), A_.nonzeroes(), M::row_wise);
> ```

If `A_reordered_` is an `optional<>` you can even use [`optional<>::emplace()`](https://en.cppreference.com/w/cpp/utility/optional/emplace) to construct the object in place:
```C++
A_reordered_.emplace(A_.N(), A_.N(), A_.nonzeroes(), M::row_wise);
```

Reply:
This worked, many thanks!